crates/bounded-executor/src/concurrent_stream.rs

```diff
@@ -0,0 +1,151 @@
+use crate::BoundedExecutor;
+use futures::{
+    stream::{self, FusedStream},
+    Future, FutureExt, Stream, StreamExt,
+};
+
+pub fn concurrent_map<St, Fut, F>(
+    stream: St,
+    executor: BoundedExecutor,
+    mut mapper: F,
+) -> impl FusedStream<Item = Fut::Output>
+where
+    St: Stream,
+    F: FnMut(St::Item) -> Fut + Send,
+    Fut: Future + Send + 'static,
+    Fut::Output: Send + 'static,
+{
+    stream
+        .flat_map_unordered(None, move |item| {
+            let future = mapper(item);
+            let executor = executor.clone();
+            stream::once(async move { executor.spawn(future).await }.boxed()).boxed()
+        })
+        .flat_map_unordered(None, |handle| {
+            stream::once(async move { handle.await.expect("result") }.boxed()).boxed()
+        })
+        .fuse()
+}
+
+#[rustversion::since(1.75)]
+pub trait ConcurrentStream: Stream {
```

Comment:
Who implements the trait?

Reply:
It's an auto trait that we apply to everyone who implements stream.